File: src/main/kotlin/cn/bakamc/folia/service/PlayerService.kt
package cn.bakamc.folia.service

import cn.bakamc.folia.db.database
import cn.bakamc.folia.db.table.*
import cn.bakamc.folia.extension.uuid
import org.bukkit.entity.Player
import org.ktorm.dsl.batchUpdate
import org.ktorm.dsl.eq
import org.ktorm.dsl.inList
import org.ktorm.entity.*
import org.ktorm.support.mysql.insertOrUpdate

object PlayerService {

    suspend fun insertOrUpdate(player: Player) {
        database {
            insertOrUpdate(PlayerInfos) {
                set(it.uuid, player.uuid)
                set(it.name, player.name)
                onDuplicateKey {
                    set(it.name, player.name)
                }
            }
        }
    }


    suspend fun getFlightEnergy(player: Player): FlightEnergy {
        return database {
            flightEnergies.find {
                val playerInfo = (it.uuid.referenceTable as PlayerInfos)
                playerInfo.uuid eq player.uuid
            } ?: run {
                val flightEnergy = FlightEnergy {
                    this.player = playerInfos.find { it.uuid eq player.uuid }!!
                    enabled = false
                    energy = 0.0
                    barVisible = true
                }
                flightEnergies.add(flightEnergy)
                flightEnergy
            }
        }
    }

    suspend fun getFlightEnergies(players: Collection<Player>): Map<Player, FlightEnergy> {
        if (players.isEmpty()) return emptyMap()
        return database {
            buildMap {
                flightEnergies.filter { flightEnergy ->
                    flightEnergy.uuid inList players.map { it.uuid }
                }.forEach {
                    this[players.find { player -> player.uuid == it.uuid }!!] = it
                }
            }
        }
    }

    suspend fun updateFlightEnergy(energy: FlightEnergy) {
        database {
            flightEnergies.update(energy)
        }
    }

    suspend fun updateFlightEnergies(flightEnergy: Collection<FlightEnergy>): Int {
        if (flightEnergy.isEmpty()) return 0
        return database {
            batchUpdate(FlightEnergies) {
                flightEnergy.forEach { e ->
                    item {
                        set(it.energy, e.energy)
                        set(it.enabled, e.enabled)
                        set(it.barVisible, e.barVisible)
                        where {
                            it.uuid eq e.uuid
                        }
                    }
                }
            }
        }.filter { it > 0 }.size
    }

}

File: bakamc-folia/src/main/scala/cn/bakamc/folia/database/services/PlayerService.scala
package cn.bakamc.folia.database.services

import cn.bakamc.folia.database.db
import cn.bakamc.folia.database.table.{FlightEnergy, flightEnergies, playerInfos}
import org.bukkit.entity.Player
import slick.jdbc.MySQLProfile.api.*

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

object PlayerService {

  def insertOrUpdate(player: Player): Future[Int] = db {
    playerInfos.insertOrUpdate(player)
  }

  def getFlightEnergy(player: Player): Future[FlightEnergy] = db {
    flightEnergies.filter(_.playerUuid === player.getUniqueId.toString)
      .result.headOption
      .flatMap {
        case Some(energy) => DBIO.successful(energy)
        case None =>
          val newEnergy = FlightEnergy(player, 0.0, false, true)
          (flightEnergies += newEnergy) >> DBIO.successful(newEnergy)
      }
  }

  def getFlightEnergies(players: Iterable[Player]): Future[Map[Player, FlightEnergy]] = {
    if (players.isEmpty) Future.successful(Map.empty)
    else {
      val playerMap = players.map(p => p.getUniqueId.toString -> p).toMap
      db {
        val playerUuids = players.map(_.getUniqueId.toString)
        flightEnergies.filter(_.playerUuid inSet playerUuids)
          .result
          .map(energies =>
            energies.map(e => playerMap(e.playerUuid) -> e).toMap
          )
      }
    }
  }

  def updateFlightEnergy(energy: FlightEnergy): Future[Int] = db {
    flightEnergies.insertOrUpdate(energy)
  }

  def updateFlightEnergies(energies: Iterable[FlightEnergy]): Future[Option[Int]] = {
    if (energies.isEmpty) Future.successful(None)
    else db {
      flightEnergies.insertOrUpdateAll(energies)
    }
  }

  def addAllPlayerEnergy(amount: Double, range: (Double, Double)): Future[Option[Int]] = {
    val (min, max) = range
    if (amount <= 0) Future.successful(None)
    else {
      val sql =
        sqlu"""
            UPDATE flight_energy
            WHERE energy <= $max
            SET energy = CASE
              WHEN energy + $amount > $max THEN $max
              WHEN energy + $amount < $min THEN $min
              ELSE energy + $amount
            END
          """
      db(sql).map(count => Some(count))
    }
  }

}